### PR TITLE
Refactor tests

### PR DESCRIPTION
**Pull request by karlding**
Refactor tests into directories, and standardize a few things in the tests. Also a few misc changes to suppress output to stderr.

I'm proposing that we use a key/value store to track the last _lucashifact_ that was sent out, as part of our ongoing enhancements to the random algorithm. This can be implemented in bash on the filesystem, and projects like [kv-bash](https://github.com/damphat/kv-bash) already do so. We can directly pull this dependency in, and then modify our random fact picker algorithm to choose a fact by performing a key lookup and ensuring that the selected fact is different than the one that was previously sent. It is trivial (relatively speaking) to implement a SQL interface on top of a key-value store, so we can avoid pulling in another dependency like SQLite.

This PR will make writing tests for integrating the key-value store a lot cleaner, and makes the tests directory more manageable to read.

### CODE REVIEW

File: src/check_imap.sh [hunk 1]
```diff
@@ -10,5 +10,5 @@ fi
 
 JSON="${RUN_DIR}/lib/JSON.sh"
 
-MATCHING=$(${JSON} < "${IMAP_SOURCE}" | grep -v "\[\]" | grep "$1")
+MATCHING=$(${JSON} < "${IMAP_SOURCE}" 2>/dev/null| grep -v "\[\]" | grep "$1")
```
Comment (Caelink) [inline src/check_imap.sh:13]: maybe generating a tmpfile log is a better solution for testing later

File: src/disrupt.sh [hunk 2]
```diff
@@ -18,7 +18,7 @@ PARTIAL="${RUN_DIR}/assets/partial"
 # Set the dice roll to zero if:
 #   - the image assets directory doesn't exist
 #   - no supported image files exist inside the image assets directory
-IMAGES_LIST=$(find "${IMG_ASSET_DIR}" -mindepth 1 -type f \( -iname '*.jpg' -o -iname '*.png' \) )
+IMAGES_LIST=$(find "${IMG_ASSET_DIR}" -mindepth 1 -type f \( -iname '*.jpg' -o -iname '*.png' \) 2>/dev/null )
```
Comment (Caelink) [inline src/disrupt.sh:21]: same as other dev null comment

File: tests/img_test/config.in [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+TEST_NAME="img_test"
+
+export EMAILS=""
+export MAIL_SCRIPT="$RUN_DIR/src/disrupt.sh"
+export FACTS_SOURCE="$TEST_DIR/$TEST_NAME/facts.lf"
```
Comment (Caelink) [inline tests/img_test/config.in:7]: Maybe we can pull the variables that are unlikely to change outside the test_name variable from another file so we don't have to change them everywhere if we ever do - like source a base_config.in file? Then you can rewrite what you care about.

File: tests/run_all.sh [hunk 4]
```diff
@@ -2,22 +2,23 @@
 
 export TEST_DIR="$PWD/tests"
 tests_to_run=(
```
Comment (Caelink) [inline tests/run_all.sh:4]: I'm personally a fan of being very explicit with tests, but we could do test discovery now.

File: tests/run_all.sh [hunk 5]
```diff
@@ -2,22 +2,23 @@
...
+    missing_img_dir_test
     url_test
 )
 
 FAILED=0
 
 for test in "${tests_to_run[@]}"; do
     echo "$test"
-    ./main.sh -r . -c "tests/${test}.in" -s "tests/${test}" | diff "tests/${test}.out" - && echo "Pass" || FAILED=1
+    ./main.sh -r . -c "tests/${test}/config.in" -s "tests/${test}/img" | diff "tests/${test}/expected.out" - && echo "Pass" || FAILED=1
```
Comment (Caelink) [inline tests/run_all.sh:21]: <3 for this, much simpler to create new tests and consistent.

**Review by Caelink — COMMENTED** [review]
Looks pretty good, put some comments. Should there be this many .gitkeep files? Seems like we don't need them anymore with the fact that there is never an empty dir in tests now.

**Comment by Caelink** [comment]
As far as the SQL wrapped key-value data store - I'm in full support of some sort of implementation which is at it's heart no more than a file in some format mapping emails to the last n facts sent (and potentially img names in the future) and rolling our own lucashi_shuf. I think it deserves its own PR as it's a pretty interesting new thing, but as long as every non-asset is still in bash and it's testable I'll support it. 

It wouldn't be too difficult to implement the SQL like commands entirely in bash without pulling in any dependencies, right? I think it'd be cool to just grab the kv-bash dependency just to make writing to a file easier and refactor image maps to use that store. From there we can definitely implement a strict subset of SQL as another script.

TL;DR go ahead and tag me on the PR, shell is love and life.

**Comment by karlding** [comment]
The `.gitkeep` files force the (empty) image directories to be tracked by `git`, so unless we're explicitly testing a missing `img` directory, I think they're still needed?

As far as the key-value store, I'm not a fan of the fact that `kv-bash` doesn't let you select a database, which would make testing a pain. There's [quinedb](https://github.com/gfredericks/quinedb) and [bash-db](https://github.com/reddec/bash-db) which is more in line with the API I was thinking about, as it would allow us to easily sub out a new database for each test.

The SQL interface would be quite a bit of work, but there are databases built on top of key-value stores (CockroachDB uses RocksDB as its backend) and they give an overview of how they do it in one of their blog posts, so the theory is sound.

File: tests/img_test/config.in (hunk 3 above)
Comment (karlding) [reply tests/img_test/config.in:7]: Ah good call.

File: src/check_imap.sh (hunk 1 above)
Comment (karlding) [reply src/check_imap.sh:13]: What do you mean? Like make this a variable that can be set in config?

Comment (Caelink) [reply src/check_imap.sh:13]: Like instead of redirecting stderr to /dev/null we could redirect all error messages to a logfile associated with the last run, a temporary file that has a ttl of like 1 day. Then we don't ignore this stuff and can look into what has gone wrong if anything ever does

Comment (Caelink) [reply src/check_imap.sh:13]: For tests specifically it should be a tmp file with a lifespan of the tests lifespan though

**Comment by Caelink** [comment]
Oh I thought you had changed the setup so they weren't in orphaned directories anymore mb. 

TBQH We don't need a very complete database if we're just tracking related images and facts that have been used for emails. Of the two you linked, bash-db source code looks better organized/cleaner. Either way as long as it's testable it's fine by me
